### PR TITLE
Installer component of Managed Cloud SDK.

### PR DESCRIPTION
**Pull request by loosebazooka**

### CODE REVIEW

File: src/main/java/com/google/cloud/tools/managedcloudsdk/Version.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk;
+
+public final class Version {
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/Version.java]: We already have CloudSdkVersion. This is a little confusing.

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/OsType.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import com.google.cloud.tools.managedcloudsdk.UnsupportedOSException;
+import java.util.Locale;
+
+/** Created by appu on 8/30/17. */
+public enum OsType {
+  MAC,
+  WINDOWS,
+  LINUX;
+
+  public static OsType getSystemOS() throws UnsupportedOSException {
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/OsType.java]: getSystemOs

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallProcessStreamHandler.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.io.InputStream;
+
+/** Listener on {@link Installer} tasks */
+public interface InstallProcessStreamHandler {
+
+  /**
+   * Use this method to display output from the install process.
+   *
+   * @param stdOut Install process standard output stream
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallProcessStreamHandler.java]: install

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/Installer.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+package com.google.cloud.tools.managedcloudsdk.internal.install;
+
+import java.nio.file.Path;
+import java.util.concurrent.Callable;
+
+/**
+ * Installer for running install scripts in a Cloud SDK download. It is a callable we use in chained
+ * asynchronous calls.
+ */
+public interface Installer extends Callable<Path> {
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/Installer.java]: What does this give us over just implementing Callable<Path> directly?

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+public final class InstallerFactory {
+
+  private final Version version;
+  private final OsType os;
+  private final boolean usageReporting;
+
+  /**
+   * Creates a new factory.
+   *
+   * @param version Version of the Cloud SDK
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java]: version

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+   * @param os The operating system of the computer this script is running on
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java]: the operating

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+   * @param usageReporting Enable or disable client side usage reporting {@code true} is enabled,
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java]: enable

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    return new NoOpInstaller(installedSdkRoot);
+  }
+
+  private InstallScriptProvider getInstallScriptProvider() {
+    switch (os) {
+      case WINDOWS:
+        return new WindowsInstallScriptProvider();
+      case MAC:
+      case LINUX:
+        return new NixInstallScriptProvider();
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java]: Nix --> Unix

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      default:
+        throw new IllegalStateException("Unexpected OSType: " + os.name());
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java]: UnsupportedOsException

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/LatestInstaller.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Implementation of {@link Installer} for 'LATEST' version cloud SDK.
+ *
+ * <p>The archive downloaded provides an install scrip that we run that downloads OS and
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/LatestInstaller.java]: script
delete "we run"

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/NixInstallScriptProvider.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk.internal.install;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** {@link InstallScriptProvider} for mac and linux. */
+public final class NixInstallScriptProvider implements InstallScriptProvider {
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/NixInstallScriptProvider.java]: UnixInstall

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/ProcessBuilderFactory.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk.internal.install;
+
+/** Simple class to abstract out the ProcessBuilder. */
+final class ProcessBuilderFactory {
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/ProcessBuilderFactory.java]: This class barely seems worth it.

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/Installer.java (hunk 4 above)
Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/Installer.java]: It is only really allowing for better documentation, by making it an `Installer` type, we can document the requirements of the call method. It is internal though, so maybe we don't need to at all. Once these 3 CLs (download/extract/install) are done, perhaps the glue class that bring them all together will make the factories and interfaces unnecessary.

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/ProcessBuilderFactory.java (hunk 12 above)
Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/ProcessBuilderFactory.java]: Its a way to make the class testable.

File: src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/LatestInstallerTest.java [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    Installer installer =
+        new LatestInstaller<>(
+            tmp.getRoot().toPath(),
+            mockInstallScriptProvider,
+            false,
+            null,
+            mockProcessBuilderFactory);
+    Path returnedSdkRoot = installer.call();
+
+    Mockito.verify(mockProcessBuilder, Mockito.times(1)).command(getExpectedCommand(false));
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/LatestInstallerTest.java]: According to the Mockito documentation, `times(1)` is used by default for `verify`.

http://static.javadoc.io/org.mockito/mockito-core/2.10.0/org/mockito/Mockito.html#4

File: src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/LatestInstallerTest.java [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+    Installer installer =
+        new LatestInstaller<>(
+            tmp.getRoot().toPath(),
+            mockInstallScriptProvider,
+            false,
+            mockHandler,
+            mockProcessBuilderFactory);
+    installer.call();
+
+    Mockito.verify(mockProcessBuilder, Mockito.times(0)).inheritIO();
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/LatestInstallerTest.java]: According to the Mockito documentation, `never()` can be used as an alias for `times(0)`.

http://static.javadoc.io/org.mockito/mockito-core/2.10.0/org/mockito/Mockito.html#4

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    this.version = version;
+    this.os = os;
+    this.usageReporting = usageReporting;
+  }
+
+  /**
+   * Returns a new {@link Installer} instance.
+   *
+   * @param installedSdkRoot path to the Cloud SDK directory
+   * @param installInstallProcessStreamHandler listener on installer script output
```
Comment (coollog) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java]: Should this be installProcessStreamHandler?

Ditto in LatestInstaller too.

File: src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactoryTest.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Parameterized.Parameter(0)
+  public OsType OS;
+
+  @Parameterized.Parameter(1)
+  public Class<? extends InstallScriptProvider> expectedInstallScriptProviderClass;
+
+  @Test
+  public void testNewInstaller_LastestVersion() {
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactoryTest.java]: Latest

**Comment by briandealwis** [comment]
I think these install scripts come from the `google-cloud-sdk` (part of the downloaded archive)?  Do we actually need to execute them?  The README notes that it's optional.

> 3. Optional. Run the install script to add Cloud SDK tools to your path, enable command-completion in your bash shell (Linux and Mac OS only) and enable usage reporting.
> 
> On Linux or Mac OS X:
> 
> ./google-cloud-sdk/install.sh
>
> On Windows:
> 
> .\google-cloud-sdk\install.bat
>
> Run install.sh --help or install.bat --help for a list of flags you can pass to this script.

File: src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    this.usageReporting = usageReporting;
+  }
+
+  /**
+   * Returns a new {@link Installer} instance.
+   *
+   * @param installedSdkRoot path to the Cloud SDK directory
+   * @param installProcessStreamHandler listener on installer script output
+   * @return a {@link LatestInstaller} instance if factory configured with 'LATEST' version or
+   *     {@link NoOpInstaller} if factory is configured with a fixed version.
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactory.java]: nit: no period

**Comment by loosebazooka** [comment]
@elharo, you make a valid point on `internal`, I'll rearrange all this with the glue CL for the installer.

File: src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 18]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (coollog) [inline src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker]: No newline at end of file

File: src/main/java/com/google/cloud/tools/managedcloudsdk/Version.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk;
+
+/** Defines the version of a Cloud SDK. */
+public final class Version {
+  public static Version LATEST = new Version();
```
Comment (coollog) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/Version.java]: This probably should be final so nothing can accidentally set the LATEST to something else.

File: src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactoryTest.java [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+        expectedInstallScriptProviderClass,
+        ((LatestInstaller) installer).getInstallScriptProvider().getClass());
+  }
+
+  @Test
+  public void testNewInstaller_fixedVersion() throws BadCloudSdkVersionException {
+    Installer installer =
+        new InstallerFactory(new Version("1.0.0"), OS, false).newInstaller(null, null);
+    Assert.assertEquals(NoOpInstaller.class, installer.getClass());
+  }
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactoryTest.java]: Maybe add a test for a bad version and check the BadCloudSdkVersionException that is thrown.

File: src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker (hunk 18 above)
Comment (loosebazooka) [reply src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker]: done

File: src/main/java/com/google/cloud/tools/managedcloudsdk/Version.java (hunk 19 above)
Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/Version.java]: done

File: src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactoryTest.java (hunk 20 above)
Comment (loosebazooka) [reply src/test/java/com/google/cloud/tools/managedcloudsdk/internal/install/InstallerFactoryTest.java]: done

**Pull request merged**